File: native/sensor-bridge/android/sensor-bridge/src/main/kotlin/com/dragonfly/sensorbridge/SyncClient.kt
package com.dragonfly.sensorbridge

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URL

/**
 * Talks to POST {apiBase}/api/glucose/sync. Wire shape lives in
 * native/sensor-bridge/shared/PROTOCOL.md.
 */
class SyncClient(
    private val apiBase: String,
    private val bridgeVersion: String = "android-0.1.0",
    private val bearerTokenProvider: () -> String? = { null },
) {

    private val json = Json {
        ignoreUnknownKeys = true
        explicitNulls = false
    }

    suspend fun uploadBatch(
        patientId: String,
        vendor: GlucoseEvent.Vendor,
        deviceName: String?,
        events: List<GlucoseEvent>,
    ): Result {
        val url = URL("${apiBase.trimEnd('/')}/api/glucose/sync")
        val payload = SyncBatch(
            patientId = patientId,
            vendor = vendor,
            deviceName = deviceName,
            bridgeVersion = bridgeVersion,
            readings = events.map { e ->
                SyncReading(
                    patientId = patientId,
                    valueMgDl = e.valueMgDl,
                    source = "cgm",
                    vendor = e.vendor ?: vendor,
                    deviceName = e.deviceName ?: deviceName,
                    context = e.context,
                    timestamp = e.timestamp,
                    trend = e.trend,
                    rawDeviceId = e.rawDeviceId,
                    readingKind = e.readingKind,
                    ingestionPath = e.ingestionPath,
                    notes = e.notes,
                )
            },
        )
        val body = json.encodeToString(payload)

        val conn = url.openConnection() as HttpURLConnection
        try {
            conn.requestMethod = "POST"
            conn.doOutput = true
            conn.setRequestProperty("Content-Type", "application/json")
            bearerTokenProvider()?.let {
                conn.setRequestProperty("Authorization", "Bearer $it")
            }
            OutputStreamWriter(conn.outputStream).use { it.write(body) }
            val code = conn.responseCode
            val text = (if (code in 200..299) conn.inputStream else conn.errorStream)
                .bufferedReader().use { it.readText() }
            if (code !in 200..299) error("sync failed ($code): $text")
            return json.decodeFromString(Result.serializer(), text)
        } finally {
            conn.disconnect()
        }
    }

    @Serializable
    data class Result(
        val accepted: Int,
        val duplicates: Int,
        val rejected: List<Rejection>,
    ) {
        @Serializable
        data class Rejection(val index: Int, val reason: String)
    }

    @Serializable
    private data class SyncBatch(
        val patientId: String,
        val vendor: GlucoseEvent.Vendor,
        val deviceName: String?,
        val bridgeVersion: String,
        val readings: List<SyncReading>,
    )

    @Serializable
    private data class SyncReading(
        val patientId: String,
        val valueMgDl: Double,
        val source: String,
        val vendor: GlucoseEvent.Vendor,
        val deviceName: String?,
        val context: GlucoseEvent.Context,
        val timestamp: String,
        val trend: GlucoseEvent.Trend? = null,
        val rawDeviceId: String? = null,
        val readingKind: GlucoseEvent.ReadingKind,
        val ingestionPath: GlucoseEvent.IngestionPath,
        val notes: String? = null,
    )
}

File: native/sensor-bridge/ios/DragonflySensorBridge/Sources/DragonflySensorBridge/SyncClient.swift
import Foundation

/// Talks to `POST {apiBase}/api/glucose/sync`. The wire shape lives in
/// `native/sensor-bridge/shared/PROTOCOL.md`.
public final class SyncClient {
    public struct Result: Decodable {
        public let accepted: Int
        public let duplicates: Int
        public let rejected: [Rejection]

        public struct Rejection: Decodable {
            public let index: Int
            public let reason: String
        }
    }

    public let apiBase: URL
    public let bridgeVersion: String
    private let session: URLSession
    private let bearerTokenProvider: () -> String?

    public init(
        apiBase: URL,
        bridgeVersion: String = "ios-0.1.0",
        bearerTokenProvider: @escaping () -> String? = { nil },
        session: URLSession = .shared
    ) {
        self.apiBase = apiBase
        self.bridgeVersion = bridgeVersion
        self.bearerTokenProvider = bearerTokenProvider
        self.session = session
    }

    public func uploadBatch(
        patientId: String,
        vendor: GlucoseEvent.Vendor,
        deviceName: String?,
        events: [GlucoseEvent]
    ) async throws -> Result {
        var url = apiBase
        url.appendPathComponent("api")
        url.appendPathComponent("glucose")
        url.appendPathComponent("sync")

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if let token = bearerTokenProvider() {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        let payload = SyncBatch(
            patientId: patientId,
            vendor: vendor,
            deviceName: deviceName,
            bridgeVersion: bridgeVersion,
            readings: events.map { event in
                SyncReading(
                    patientId: patientId,
                    valueMgDl: event.valueMgDl,
                    source: "cgm",
                    vendor: event.vendor ?? vendor,
                    deviceName: event.deviceName ?? deviceName,
                    context: event.context,
                    timestamp: ISO8601DateFormatter().string(from: event.timestamp),
                    trend: event.trend,
                    rawDeviceId: event.rawDeviceId,
                    readingKind: event.readingKind,
                    ingestionPath: event.ingestionPath,
                    notes: event.notes
                )
            }
        )

        // Wire keys are camelCase to match packages/shared/src/types.ts
        // and the Hono validator. Do not convert to snake_case.
        request.httpBody = try JSONEncoder().encode(payload)

        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw URLError(.badServerResponse)
        }
        guard (200..<300).contains(http.statusCode) else {
            throw NSError(domain: "DragonflySensorBridge.SyncClient", code: http.statusCode, userInfo: [
                NSLocalizedDescriptionKey: String(data: data, encoding: .utf8) ?? "sync failed"
            ])
        }
        return try JSONDecoder().decode(Result.self, from: data)
    }

    private struct SyncBatch: Encodable {
        let patientId: String
        let vendor: GlucoseEvent.Vendor
        let deviceName: String?
        let bridgeVersion: String
        let readings: [SyncReading]
    }

    private struct SyncReading: Encodable {
        let patientId: String
        let valueMgDl: Double
        let source: String
        let vendor: GlucoseEvent.Vendor
        let deviceName: String?
        let context: GlucoseEvent.Context
        let timestamp: String
        let trend: GlucoseEvent.Trend?
        let rawDeviceId: String?
        let readingKind: GlucoseEvent.ReadingKind
        let ingestionPath: GlucoseEvent.IngestionPath
        let notes: String?
    }
}
